crates/runtime/src/accelerated_table/refresh.rs

```diff
@@ -0,0 +1,683 @@
+use datafusion::execution::config::SessionConfig;
+use datafusion::logical_expr::Operator;
+use datafusion::physical_plan::{collect, ExecutionPlanProperties};
+use datafusion::{
+    datasource::TableProvider, execution::context::SessionContext, logical_expr::Expr,
+};
+use futures::Stream;
+use futures::{stream::BoxStream, StreamExt};
+use snafu::prelude::*;
+use spicepod::component::dataset::acceleration::RefreshMode;
+
+use spicepod::component::dataset::TimeFormat;
+use tokio::sync::mpsc::Receiver;
+use tokio::sync::oneshot;
+use tokio::sync::RwLock;
+use tokio_stream::wrappers::ReceiverStream;
+
+use crate::datafusion::filter_converter::TimestampFilterConvert;
+use crate::object_store_registry::default_runtime_env;
+use crate::{
+    dataconnector::get_data,
+    dataupdate::{DataUpdate, DataUpdateExecutionPlan, UpdateType},
+    status,
+    timing::TimeMeasurement,
+};
+
+#[derive(Clone, Debug)]
+pub struct Refresh {
+    pub(crate) time_column: Option<String>,
+    pub(crate) time_format: Option<TimeFormat>,
+    pub(crate) check_interval: Option<Duration>,
+    pub(crate) sql: Option<String>,
+    pub(crate) mode: RefreshMode,
+    pub(crate) period: Option<Duration>,
+}
+
+impl Refresh {
+    #[allow(clippy::needless_pass_by_value)]
+    #[must_use]
+    pub(crate) fn new(
+        time_column: Option<String>,
+        time_format: Option<TimeFormat>,
+        check_interval: Option<Duration>,
+        sql: Option<String>,
+        mode: RefreshMode,
+        period: Option<Duration>,
+    ) -> Self {
+        Self {
+            time_column,
+            time_format,
+            check_interval,
+            sql,
+            mode,
+            period,
+        }
+    }
+}
+
+pub(crate) enum AccelerationRefreshMode {
+    Full(Receiver<()>),
+    Append(Option<Receiver<()>>),
+}
+
+pub(crate) struct Refresher {
+    dataset_name: String,
+    federated: Arc<dyn TableProvider>,
+    refresh: Arc<RwLock<Refresh>>,
+    accelerator: Arc<dyn TableProvider>,
+}
+
+impl Refresher {
+    pub(crate) fn new(
+        dataset_name: String,
+        federated: Arc<dyn TableProvider>,
+        refresh: Arc<RwLock<Refresh>>,
+        accelerator: Arc<dyn TableProvider>,
+    ) -> Self {
+        Self {
+            dataset_name,
+            federated,
+            refresh,
+            accelerator,
+        }
+    }
+
+    pub(crate) async fn start(
+        &self,
+        acceleration_refresh_mode: AccelerationRefreshMode,
+        ready_sender: oneshot::Sender<()>,
+    ) {
+        let dataset_name = self.dataset_name.clone();
+        let mut stream = self.stream_updates(acceleration_refresh_mode).await;
+
+        let ctx = SessionContext::new();
+
+        let mut ready_sender = Some(ready_sender);
+
+        loop {
+            let future_result = stream.next().await;
+
+            match future_result {
+                Some(data_update) => {
+                    let Ok(data_update) = data_update else {
```

Comment:
I think the inverse is clearer (i.e. continue on `Err(e)`).

Reply:
The reason of leaving so is we will need to do more to mutate the data if it's needed (remove duplication within the new data, inbetween new and old data).